### PR TITLE
[SYCLomatic] Support migration of vset/vset2/vset4 instructions

### PR DESCRIPTION
**Pull request by yihwang-nv**
Support migration of inline PTX vset/vset2/vset4 instructions.
The corresponding test PR for these additions is: https://github.com/oneapi-src/SYCLomatic-test/pull/663

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 1]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
   return detail::extend_binary<RetT, true>(a, b, c, maximum(), second_op);
 }
 
+/// Extend \p a and \p b to 33 bit and compare input values using specified
+/// comparison \p cmp .
+///
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1326]: Need a empty comment line?

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 2]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
   return detail::extend_vbinary2<RetT, true, false>(a, b, c, maximum());
 }
 
+/// Extend \p a and \p b to 33 bit and vectorized compare input values using
+///
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1762]: Need delete.

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 3]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] b The second value
+/// \returns The comparison result of the two extended values.
+template <typename AT, typename BT, typename BinaryOperation>
+inline constexpr unsigned extend_compare(AT a, BT b, BinaryOperation cmp) {
+  return detail::extend_binary<unsigned, false>(a, b, cmp);
+}
+
+/// Extend Inputs to 33 bit, and compare input values using specified comparison
+/// \p cmp , then do \p second_op with \p c .
+///
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1341]: Same

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 4]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
 
+/// Extend \p a and \p b to 33 bit and compare input values using specified
+/// comparison \p cmp .
+///
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \returns The comparison result of the two extended values.
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1333]: Miss a `\param [in] cmp`

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 5]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \p cmp , then do \p second_op with \p c .
+///
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \tparam [in] SecondBinaryOperation The type of the second operation
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \param [in] c The third value
+/// \param [in] second_op The operation to do with the third value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1350]: Miss a `\param [in] cmp`

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 6]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \returns The comparison result of the two extended values. and \p second_op
+/// with \p c .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1352]:
```suggestion
/// with \p c.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 7]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
 }
 
+/// Extend \p a and \p b to 33 bit and vectorized compare input values using
+///
+/// specified comparison \p cmp .
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \param [in] a The first value
+/// \param [in] b The second value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1768]: Miss a `\param [in] cmp`

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 8]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \returns The comparison result of the two extended values.
+template <typename AT, typename BT, typename BinaryOperation>
+inline constexpr unsigned extend_vcompare2(AT a, BT b, BinaryOperation cmp) {
+  return detail::extend_vbinary2<unsigned, false, false>(a, b, 0, cmp);
+}
+
+/// Extend Inputs to 33 bit, and vectorized compare input values using specified
+/// comparison \p cmp , then do \p second_op with \p c .
+///
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1778]: Need a empty comment line?

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 9]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \returns The comparison result of the two extended values.
+template <typename AT, typename BT, typename BinaryOperation>
+inline constexpr unsigned extend_vcompare2(AT a, BT b, BinaryOperation cmp) {
+  return detail::extend_vbinary2<unsigned, false, false>(a, b, 0, cmp);
+}
+
+/// Extend Inputs to 33 bit, and vectorized compare input values using specified
+/// comparison \p cmp , then do \p second_op with \p c .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// comparison \p cmp, then do \p second_op with \p c.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 10]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
   return detail::extend_binary<RetT, true>(a, b, c, maximum(), second_op);
 }
 
+/// Extend \p a and \p b to 33 bit and compare input values using specified
+/// comparison \p cmp .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1325]:
```suggestion
/// comparison \p cmp.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 11]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \returns The comparison result of the two extended values.
+template <typename AT, typename BT, typename BinaryOperation>
+inline constexpr unsigned extend_compare(AT a, BT b, BinaryOperation cmp) {
+  return detail::extend_binary<unsigned, false>(a, b, cmp);
+}
+
+/// Extend Inputs to 33 bit, and compare input values using specified comparison
+/// \p cmp , then do \p second_op with \p c .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1340]:
```suggestion
/// \p cmp, then do \p second_op with \p c.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 12]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
   return detail::extend_vbinary2<RetT, true, false>(a, b, c, maximum());
 }
 
+/// Extend \p a and \p b to 33 bit and vectorized compare input values using
+///
+/// specified comparison \p cmp .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1763]:
```suggestion
/// specified comparison \p cmp.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 13]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// comparison \p cmp , then do \p second_op with \p c .
+///
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \tparam [in] SecondBinaryOperation The type of the second operation
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \param [in] c The third value
+/// \param [in] second_op The operation to do with the third value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: The comment is wrong.

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 14]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+  return detail::extend_vbinary2<unsigned, false, false>(a, b, 0, cmp);
+}
+
+/// Extend Inputs to 33 bit, and vectorized compare input values using specified
+/// comparison \p cmp , then do \p second_op with \p c .
+///
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \tparam [in] SecondBinaryOperation The type of the second operation
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: Not have this param.

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 15]
```diff
@@ -1721,6 +1756,39 @@ inline constexpr RetT extend_vmax2_sat(AT a, BT b, RetT c) {
...
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \param [in] c The third value
+/// \param [in] second_op The operation to do with the third value
+/// \returns The comparison result of the two extended values. and \p second_op
+/// with \p c .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// with \p c.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 16]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
   return detail::extend_vbinary4<RetT, true, false>(a, b, c, maximum());
 }
 
+/// Extend \p a and \p b to 33 bit and vectorized compare input values using
+///
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:2059]: Need delete

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 17]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+/// specified comparison \p cmp .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:2060]:
```suggestion
/// specified comparison \p cmp.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 18]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \returns The comparison result of the two extended values.
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:2067]: Miss a `\param [in] cmp`

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 19]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+template <typename AT, typename BT, typename BinaryOperation>
+inline constexpr unsigned extend_vcompare4(AT a, BT b, BinaryOperation cmp) {
+  return detail::extend_vbinary4<unsigned, false, false>(a, b, 0, cmp);
+}
+
+/// Extend Inputs to 33 bit, and vectorized compare input values using specified
+/// comparison \p cmp , then do \p second_op with \p c .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// comparison \p cmp, then do \p second_op with \p c.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 20]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+///
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \tparam [in] SecondBinaryOperation The type of the second operation
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: Need not.

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 21]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \param [in] c The third value
+/// \param [in] second_op The operation to do with the third value
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: wrong param.

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 22]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+/// \returns The comparison result of the two extended values. and \p second_op
+/// with \p c .
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]:
```suggestion
/// with \p c.
```

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 23]
```diff
@@ -1985,6 +2053,39 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+/// \param [in] b The second value
+/// \returns The comparison result of the two extended values.
+template <typename AT, typename BT, typename BinaryOperation>
+inline constexpr unsigned extend_vcompare4(AT a, BT b, BinaryOperation cmp) {
+  return detail::extend_vbinary4<unsigned, false, false>(a, b, 0, cmp);
+}
+
+/// Extend Inputs to 33 bit, and vectorized compare input values using specified
+/// comparison \p cmp , then do \p second_op with \p c .
+///
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:2075]: Need delete.

**Review by tangjj11 — CHANGES_REQUESTED** [review]

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 1 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp:1326]: Without this blank line, clang-format will output a incorrect format.
https://github.com/oneapi-src/SYCLomatic/blob/ba127eb6ed0c83348ff0330e6ddac3ab71b6170d/clang/include/clang/Analysis/CloneDetection.h#L49-L62

**Review by tangjj11 — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 24]
```diff
@@ -1321,6 +1321,41 @@ inline constexpr RetT extend_max_sat(AT a, BT b, CT c,
...
 }
 
+/// Extend \p a and \p b to 33 bit and compare input values using specified
+/// comparison \p cmp .
+///
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \param [in] a The first value
+/// \param [in] b The second value
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1331]: no introduce for the cmp argument.

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 5 above)
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/math.hpp:1350]: no introduce on cmp argument?

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 13 above)
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: no second_op in this function

File: clang/runtime/dpct-rt/include/dpct/math.hpp (hunk 20 above)
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: no SecondBinaryOperation in this function.

File: clang/runtime/dpct-rt/include/dpct/math.hpp [hunk 25]
```diff
@@ -1985,6 +2056,40 @@ inline constexpr RetT extend_vmax4_sat(AT a, BT b, RetT c) {
...
+  return detail::extend_vbinary4<unsigned, false, false>(a, b, 0, cmp);
+}
+
+/// Extend Inputs to 33 bit, and vectorized compare input values using specified
+/// comparison \p cmp , then add the result with \p c .
+///
+/// \tparam [in] AT The type of the first value, can only be 32 bit integer
+/// \tparam [in] BT The type of the second value, can only be 32 bit integer
+/// \tparam [in] BinaryOperation The type of the compare operation
+/// \tparam [in] SecondBinaryOperation The type of the second operation
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/math.hpp]: L 2080 is not dead comments.

Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/math.hpp]: removed

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**